app/models/order.rb

```diff
@@ -1,12 +1,59 @@
 class Order < ApplicationRecord
+  attr_accessor :credit_card_id
+
+  validates :state, presence: true
+  belongs_to :user
+  has_many :order_items
+  has_one :payment
+
   self.state_machine({
-    cart: [:placed, :canceled],
+    cart: [:placed],
     placed: [:canceled],
     canceled: []
-  })
+   })
 
-  validates :state, presence: true
-  belongs_to :user
-  has_many :order_items, as: :source
-  #todo implement state transition logic
+  before_transition_to :placed do
+    if order_items.to_a.sum(&:quantity) == 0
+      errors[:base] << "Order must contain at least one item"
+    end
+
+    order_items.each do |item|
+      if item.source.on_hand_count < item.quantity
+        errors[:base] << "There is not enough #{item.name} in inventory to complete your order."
+      end
+    end
+
+    #consume the orders nom nom nom <(-.-<)
+    order_items.each do |item|
+      item.source.on_hand_count -= item.quantity
+      item.source.available = item.source.on_hand_count > 0
```

Comment:
available is intended for administrators to make products show up for sale or not. Doesn't need to be updated here.